File: leetcode/残酷刷题/1882. Process Tasks Using Servers.cpp
#include <bits/stdc++.h>
using namespace std;
/*
空闲服务器队列，按照 weight, idx 排序，每次分配给一个任务之后，进入繁忙服务器队列
繁忙服务器队列，按照结束时间排序，结束之后进入空闲服务器队列，因为最后多出一些任务的时候需要等待繁忙服务器，所以结束时间相同也需要按照 weight 排序
*/
typedef array<int, 2> AI2;
typedef array<int, 3> AI3;
class Solution {
   public:
    vector<int> assignTasks(vector<int>& servers, vector<int>& tasks) {
        vector<int> ans(tasks.size());
        priority_queue<AI2, vector<AI2>, greater<>> free;
        priority_queue<AI3, vector<AI3>, greater<>> busy;
        // 将所有服务器放入空闲队列
        for (int i = 0; i < servers.size(); i++) {
            // wight, idx
            free.push({servers[i], i});
        }
        queue<int> jobs;
        // 处理所有的 tasks 放入任务队列中
        for (int j = 0; j < tasks.size(); j++) {
            // 看看哪些服务器变成空闲了
            while (busy.size() && busy.top()[0] <= j) {
                auto b = busy.top();
                busy.pop();
                free.push({b[1], b[2]});
            }
            // 放入 jobs 队列
            jobs.push(j);
            // 如果有空闲服务器可以使用的话，一个一个执行 jobs 的任务
            while (free.size() && jobs.size()) {
                auto job = jobs.front();
                jobs.pop();
                auto f = free.top();
                free.pop();
                // endTime, idx
                busy.push({j + tasks[job], f[0], f[1]});
                ans[job] = f[1];
            }
        }
        // 可能有在 n-1 时间内仍未执行完的任务
        // 这个时候 free 队列一定为空，否则 jobs 会在之前已经执行了
        // 直接看 busy 队列即可，每次拿一个出来，直接跳到 server 的空闲时间
        while (jobs.size()) {
            assert(free.empty());
            auto j = jobs.front();
            jobs.pop();
            auto b = busy.top();
            busy.pop();
            busy.push({b[0] + tasks[j], b[1], b[2]});
            ans[j] = b[2];
        }
        return ans;
    }
};

File: leetcode/残酷刷题/1882. Process Tasks Using Servers.ts
export {}
class Heap<T=number> {
  data: Array<T | null>
  lt: (i: number, j: number) => boolean
  constructor ()
  constructor (data: T[])
  constructor (cmp: (lhs: T, rhs: T) => boolean)
  constructor (data: T[], cmp: (lhs: T, rhs: T) => boolean)
  constructor (data: (T[] | ((lhs: T, rhs: T) => boolean)), cmp: (lhs: T, rhs: T) => boolean)
  constructor (data: (T[] | ((lhs: T, rhs: T) => boolean)) = [], cmp = (lhs: T, rhs: T) => lhs < rhs) {
    if (typeof data === 'function') {
      cmp = data
      data = []
    }
    this.data = [null, ...data]
    this.lt = (i, j) => cmp(this.data[i]!, this.data[j]!)
    for (let i = this.size(); i > 0; i--) this.heapify(i)
  }

  size (): number {
    return this.data.length - 1
  }

  push (v: T): void {
    this.data.push(v)
    let i = this.size()
    while ((i >> 1 !== 0) && this.lt(i, i >> 1)) this.swap(i, i >>= 1)
  }

  pop (): T {
    this.swap(1, this.size())
    const top = this.data.pop()
    this.heapify(1)
    return top!
  }

  top (): T { return this.data[1]! }
  heapify (i: number): void {
    while (true) {
      let min = i
      const [l, r, n] = [i * 2, i * 2 + 1, this.data.length]
      if (l < n && this.lt(l, min)) min = l
      if (r < n && this.lt(r, min)) min = r
      if (min !== i) {
        this.swap(i, min); i = min
      } else break
    }
  }

  swap (i: number, j: number): void {
    const d = this.data;
    [d[i], d[j]] = [d[j], d[i]]
  }
}

function assignTasks (servers: number[], tasks: number[]): number[] {
  const ans = []
  const free = new Heap<{
    weight: number
    index: number
  }>((a, b) => {
    if (a.weight !== b.weight) return a.weight < b.weight
    return a.index < b.index
  })
  const busy = new Heap<{
    endTime: number
    weight: number
    index: number
  }>((a, b) => {
    if (a.endTime !== b.endTime) return a.endTime < b.endTime
    if (a.weight !== b.weight) return a.weight < b.weight
    return a.index < b.index
  })

  for (let i = 0; i < servers.length; i++) {
    free.push({ weight: servers[i], index: i })
  };
  const jobs = []
  for (let i = 0; i < tasks.length; i++) {
    while (busy.size() && busy.top().endTime <= i) {
      const { weight, index } = busy.pop()
      free.push({
        weight,
        index
      })
    }
    jobs.push(i)
    while (free.size() && jobs.length) {
      const job = jobs.shift()!
      const { weight, index } = free.pop()
      busy.push({
        endTime: i + tasks[job],
        weight,
        index
      })
      ans[job] = index
    }
  }
  while (jobs.length) {
    const job = jobs.shift()!
    const { endTime, weight, index } = busy.pop()
    ans[job] = index
    busy.push({ endTime: endTime + tasks[job], weight, index })
  }
  return ans
}
